src/overlays/actors/ovl_En_Slime/z_en_slime.c

```diff
@@ -88,103 +128,1169 @@ static DamageTable D_80A31B1C = {
+        this->actor.velocity.y = 0.0f;
+    }
+    Actor_SetScale(&this->actor, 0.01f);
+    this->actionFunc = EnSlime_Expire;
+    this->actor.speed = 0.0f;
+}
+
+/**
+ * Spawn random effects (bubbles, splash) for death sequence.
+ * If chuchu is red, green, or yellow, spawn item drop.
+ * Shrink actor scale to 0.
+ * Call HideDead function for cleanup.
+ */
+void EnSlime_Expire(EnSlime* this, PlayState* play) {
+    s32 i;
+    Vec3f velocity;
+    Vec3f effectPos;
+    f32 random;
+    s32 factor;
+
+    if (Math_StepToF(&this->actor.scale.x, 0.0f, 0.002f)) {
+        effectPos.x = this->actor.world.pos.x;
+        effectPos.y = this->actor.world.pos.y + 20.0f;
+        effectPos.z = this->actor.world.pos.z;
+        EffectSsGSplash_Spawn(play, &effectPos, NULL, NULL, 1, 800);
+
+        for (i = 0; i < 15; i++) {
+            random = Rand_ZeroOne();
+
+            if (Rand_ZeroOne() < 0.5f) {
+                factor = -1;
+            } else {
+                factor = 1;
+            }
+
+            velocity.x = factor * ((random * 2.5f) + 2.0f);
+
+            random = Rand_ZeroOne();
+            if (Rand_ZeroOne() < 0.5f) {
+                factor = -1;
+            } else {
+                factor = 1;
+            }
+
+            velocity.z = factor * ((random * 2.5f) + 2.0f);
+
+            velocity.y = (Rand_ZeroOne() * 6.0f) + 2.0f;
+            EffectSsDtBubble_SpawnCustomColor(play, &effectPos, &velocity, &sBubbleAccel, &sBubblePrimColor,
+                                              &sBubbleEnvColor, Rand_S16Offset(40, 20), 20, 0);
+        }
+
+        if (this->actor.params == EN_SLIME_TYPE_YELLOW) {
+            Item_DropCollectible(play, &this->actor.world.pos, ITEM00_ARROWS_10);
+        } else if (this->actor.params == EN_SLIME_TYPE_GREEN) {
+            Item_DropCollectible(play, &this->actor.world.pos, ITEM00_MAGIC_SMALL);
+        } else if (this->actor.params == EN_SLIME_TYPE_RED) {
+            Item_DropCollectible(play, &this->actor.world.pos, ITEM00_RECOVERY_HEART);
+        }
+        EnSlime_HideDead(this);
+    }
+    this->actor.scale.y = this->actor.scale.x;
+    this->actor.scale.z = this->actor.scale.x;
+}
+
+/**
+ * Calculate a new "current" position value that is a multiple of 60
+ * away from the "home" position.
+ * This is for use by blue chuchus frozen into an ice block so that
+ * their position is a multiple of the ice block size.
+ */
+f32 EnSlime_SnapIceBlockPosition(f32 currentPosition, f32 homePosition) {
+    s32 dist = currentPosition - homePosition;
+
+    if (dist > 0) {
+        dist += 30;
+    } else {
+        dist -= 30;
+    }
+
+    return ((dist / 60) * 60.0f) + homePosition;
+}
+
+/**
+ * Prepare actor for ice block to spawn around it.
+ * The actor will stop moving, have AC_ON cleared, and
+ * target position will be snapped to a multiple of the size of
+ * an ice block.
+ * Set action function to SpawnIceBlock
+ */
+void EnSlime_InitIceBlock(EnSlime* this) {
```

Comment:
Another setup function, I would call this one `EnSlime_SetupSpawnIceBlock` and keep the name for the one below.